httpx/_urls.py

```diff
@@ -507,21 +365,33 @@ def __eq__(self, other: typing.Any) -> bool:
         return isinstance(other, (URL, str)) and str(self) == str(URL(other))
 
     def __str__(self) -> str:
-        return self._uri_reference.unsplit()
+        return str(self._uri_reference)
 
     def __repr__(self) -> str:
-        class_name = self.__class__.__name__
-        url_str = str(self)
-        if self._uri_reference.userinfo:
-            # Mask any password component in the URL representation, to lower the
-            # risk of unintended leakage, such as in debug information and logging.
-            username = quote(self.username)
-            url_str = (
-                rfc3986.urlparse(url_str)
-                .copy_with(userinfo=f"{username}:[secure]")
-                .unsplit()
-            )
-        return f"{class_name}({url_str!r})"
+        scheme, userinfo, host, port, path, query, fragment = self._uri_reference
+
+        if ":" in userinfo:
+            # Mask any password component.
+            userinfo = f'{userinfo.split(":")[0]}:[secure]'
+
+        authority = "".join(
+            [
+                f"{userinfo}@" if userinfo else "",
+                f"[{host}]" if ":" in host else host,
+                f":{port}" if port is not None else "",
+            ]
+        )
+        url = "".join(
```

Comment:
Question: could this just be str(self._url_reference)? Just curious what the implications are of the different rendering methods are. I think it'd be more parsimonious to have a single URL->str but I could see this being a backwards-compat thing.

Reply:
Right now we're doing this here because we need to mask the password for `__repr__`, but not for the `__str__`.

We could perhaps simplify a little by having a `to_string(mask_password: bool)` method on the `ParseResult` class.

Reply:
Ohhhh right, because of this, right? https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1

> We could perhaps simplify a little by having a to_string(mask_password: bool) method on the ParseResult class.

That also sounds reasonable, if it were my codebase, I'd prefer that. Not a big deal either way.

Reply:
> I think it would be better if it didn't use square braces to elide the password, since that screws up later parses, but that's beyond the purview of this PR

Hrm, yes, we could certainly reconsider that. As you say tho, we'd do so independently of this pull request.